jbi/queue.py

```diff
@@ -0,0 +1,315 @@
+            order by the timestamp of the payload event.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def size(self) -> int:
+        pass
+
+
+class MemoryBackend(QueueBackend):
+    def __init__(self):
+        self.existing: dict[int, list[QueueItem]] = defaultdict(list)
+
+    def ping(self):
+        return True
+
+    async def clear(self):
+        logger.debug("Clearing queue")
+        self.existing.clear()
+
+    async def put(self, item: QueueItem):
+        bisect.insort_right(
+            self.existing[item.payload.bug.id],
+            item,
+            key=lambda i: i.payload.event.time or 0,
+        )
+
+    async def get(self, bug_id: int) -> list[QueueItem]:
+        # though we're using a defaultdict, check if the bug is in the queue
+        # this way so that we don't create a key for a bug with no items
+        if bug_id not in self.existing:
+            return []
+        items = self.existing[bug_id]
+        if not items:
+            logger.warn("No items for bug %s, but present in queue", bug_id)
+        return items
+
+    async def get_all(self) -> dict[int, list[QueueItem]]:
+        empty_entries = [not items for items in self.existing.values]
+        if empty_entries:
+            logger.warn(
+                "No items for bugs %s, but present in queue", ",".join(empty_entries)
+            )
+        return self.existing
+
+    async def remove(self, bug_id: int, identifier: str):
+        filtered_items = [
+            i for i in self.existing[bug_id] if i.identifier != identifier
+        ]
+        if not len(filtered_items):
+            del self.existing[bug_id]
+            logger.debug(
+                "Removed %s and entry for bug %s from queue", identifier, bug_id
+            )
+        else:
+            self.existing[bug_id] = filtered_items
+            logger.debug("Removed %s from queue for bug %s", identifier, bug_id)
+
+    @property
+    def size(self) -> int:
+        return sum(len(v) for v in self.existing.values())
+
+
+class FileBackend(QueueBackend):
+    def __init__(self, location):
+        self.location = Path(location)
+        self.location.mkdir(parents=True, exist_ok=True)
+
+    def ping(self):
+        try:
+            with tempfile.TemporaryDirectory(dir=self.location) as temp_dir:
+                with tempfile.TemporaryFile(dir=temp_dir) as f:
+                    f.write(b"")
+            return True
+        except Exception:
+            logger.exception("Could not write to file backed queue")
+            return False
+
+    async def clear(self):
+        shutil.rmtree(self.location)
+
+    async def put(self, item: QueueItem):
+        folder = self.location / f"{item.payload.bug.id}"
+        folder.mkdir(exist_ok=True)
+        path = folder / (item.identifier + ".json")
+        path.write_text(item.model_dump_json())
+        logger.debug("%d items in dead letter queue", self.size)
```

Comment:
self.size will be executed regardless of log level, correct? I don't think python does any magic here to not do that.

In that case, this may be problematic with a busy FileBackend. But we _shouldn't_ have that.

Reply:
> self.size will be executed regardless of log level, correct?

Nope, size will be lazily computed for this log line only when the log level is set to debug